### PR TITLE
refactor setTargetCPU()

### PR DESCRIPTION
**Pull request by WalterBright**
There is a simpler function lurking within it.

### CODE REVIEW

**Comment by WalterBright** [comment]
This relies on https://github.com/dlang/dmd/pull/8663

**Comment by jacob-carlborg** [comment]
#8663 has been merged, please rebase.

**Comment by WalterBright** [comment]
rebased

**Review by Geod24 — CHANGES_REQUESTED** [review]
Personally I don't find this to be much of an enhancement. Instead of passing `CPU` to `setTargetCPU` we are passing the whole `Param` struct, which contains just about everything passed on the command line.

So as an external person reads the code, it is not clear what this function does at call site. It **could** do about anything: change any value in `Param`, or read any value in `Param`.

On the other hand, when looking at the call site, the current code clearly shows that only `params.cpu` is read, and `params.cpu` is written, and the `setTargetCPU` does not have to know about `Params`.

**Comment by WalterBright** [comment]
> does not have to know about Params

This is incorrect, .cpu, .is64bit and .isOSX fields are read:
```
if (params.is64bit || params.isOSX)
```
Another problem with the original is its use of TARGET.OSX instead of params.isOSX, which is a bit schizophrenic because of the params.is64bit.

A close read of the original shows it takes four parameters:

1. cpu
2. params.is64bit
3. TARGET.OSX
4. cpuid

There is no indication of the latter three in the parameter list for the function. A case could be made for passing cpuid in as an argument, too.

Ideally, params should be passed in as `const ref`, but there's a problem with the design that there should be a globals.cpu to set instead of params.cpu, but that's a refactoring for another day.

It took me several reads of the original tangled logic to figure it all out. I like the new layout much better.

File: src/dmd/mars.d [hunk 1]
```diff
@@ -1474,51 +1475,38 @@ extern(C) void printGlobalConfigs(FILE* stream)
...
- *      value to generate code for
+ *      params = parameters set by command line switch
  */
 
-private CPU setTargetCPU(CPU cpu)
+private void setTargetCPU(ref Param params)
 {
-    // Determine base line for target
-    CPU baseline = CPU.x87;
-    if (global.params.is64bit)
```
Comment (Geod24) [inline src/dmd/mars.d]: Right, I missed this

**Review by Geod24 — APPROVED** [review]

**Comment by Geod24** [comment]
Thanks for the breakdown, it does make sense

**Comment by jacob-carlborg** [comment]
Isn't better to have the function return the value instead of setting it inside the function?

**Pull request merged**